cli/functionary/client.py

```diff
@@ -0,0 +1,104 @@
+import json
+
+import click
+import requests
+
+from .config import get_config_value
+
+
+def get(endpoint):
+    """
+    Gets any data associated with an endpoint from the api
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+
+    Returns:
+        Data: Data from endpoint as Python list/dict
+
+    """
+    response = _send_request(endpoint, "get")
+    data = json.loads(response.text).get("results")
+    return data
+
+
+def post(endpoint, data=None, files=None):
+    """
+    Post provides data or files to endpoint
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+        Data: Any data to put in the request's data field
+        Files: Any files to put in the request's files field
+
+    Returns:
+        Data: Response from endpoint as Python list/dict
+
+    """
+    response = _send_request(endpoint, "post", post_data=data, post_files=files)
+    data = json.loads(response.text)
+    return data
+
+
+def _send_request(endpoint, request_type, post_data=None, post_files=None):
+    """
+    Helper function for get and post that sends the request and handles any errors
+    that arise
+
+    Args:
+        Endpoint: the name of the endpoint to get data from
+        Request_type: Either post or get
+        Post_data: Any data to put in the post request's data field
+        Post_files: Any files to put in the post request's files field
+
+    Returns:
+        Response: A Response object generated from the request
+
+    Raises:
+        ClickException: Raised if cannot connect to host, permission issue
+        exists, user has not set a required field, or other request failure
+
+    """
+    host = get_config_value("host")
+    url = host + f"/api/v1/{endpoint}"
+    try:
+        if endpoint == "api-token-auth" and request_type == "post":
```

Comment:
There shouldn't be any checks for specific endpoints within this code.  There are two possibilities here:

* The call to the api-token-auth endpoint has some completely unique need that nothing else going through this code path might share.  If that is the case, it would make more sense to just have the spot where api-token-auth is called not use this function and instead just call `requests.post` directly.
* Or, the more likely scenario, the api-token-auth call is indicative of a more general use case that this function could be reworked to accommodate.

Let's say it's that second one and that the general use case is calls to endpoints where authorization is not required.  That is, calls that you can make or are expected to make without having first logged in.  The act of logging in itself is the situation you're dealing with here, but there could conceivably be others.

You could handle this by having this (and the `get` and `post` functions that call it) take an argument `authorization` which defaults to `True`, but in the login scenario you would set to `False`.  Then that would drive whether or not you set the authorization header here.

Having said all that, the first route to try is just always sending the authorization token if you have it and not sending it if you don't.  In the login scenario, if the user already has an authorization token and is trying to login (possibly under a different account), it's okay to send that old token because it's just going to be ignored by that endpoint anyway.  For other endpoints, if the token was required and you don't send it, the response from the API is going to tell you that.